### PR TITLE
executor: support cleanup dangling index data command

### PR DESCRIPTION
**Pull request by jackysp**
Support `admin cleanup index table_name index_name` to clean up the dangling index.
PTAL @zimulala @coocood

### CODE REVIEW

File: parser/parser.y [hunk 1]
```diff
@@ -260,6 +260,7 @@ import (
 	cascaded	"CASCADED"
 	charsetKwd	"CHARSET"
 	checksum	"CHECKSUM"
+	cleanup     "CLEANUP"
```
Comment (zimulala) [inline parser/parser.y]: Keep alignment.

File: parser/parser.y [hunk 2]
```diff
@@ -4782,6 +4783,14 @@ AdminStmt:
 			Index: string($5),
 		}
 	}
+|	"ADMIN" "CLEANUP" "INDEX" TableName Identifier
```
Comment (zimulala) [inline parser/parser.y]: Do we need to add tests in parser_test.go ?

File: executor/admin.go [hunk 3]
```diff
@@ -457,3 +459,165 @@ func (e *RecoverIndexExec) NextChunk(ctx context.Context, chk *chunk.Chunk) erro
...
+	idxResult        distsql.SelectResult
+
+	//tblChunk         *chunk.Chunk
+	//tblResult        distsql.SelectResult
+
+	matchedIndex     map[int64]idxData
+}
+
+type idxData struct {
+	matched bool
```
Comment (zimulala) [inline executor/admin.go]: It seems we don't use this field.

File: executor/admin.go [hunk 4]
```diff
@@ -457,3 +459,165 @@ func (e *RecoverIndexExec) NextChunk(ctx context.Context, chk *chunk.Chunk) erro
...
+}
+
+// NextChunk implements the Executor NextChunk interface.
+func (e *CleanupIndexExec) NextChunk(ctx context.Context, chk *chunk.Chunk) error {
+	chk.Reset()
+	if e.done {
+		return nil
+	}
+	handleIdx := len(e.idxCols) - 1
+	errInTxn := kv.RunInNewTxn(e.ctx.GetStore(), true, func(txn kv.Transaction) error {
```
Comment (zimulala) [inline executor/admin.go]: Do we need to handle this code in batch? I'm afraid that we need to clean up a lot of data. At the same time, some people also modified the data. Then we need to redo it.

**Comment by jackysp** [comment]
PTAL @coocood @zimulala

**Comment by coocood** [comment]
LGTM

File: executor/admin.go [hunk 5]
```diff
@@ -457,3 +459,247 @@ func (e *RecoverIndexExec) NextChunk(ctx context.Context, chk *chunk.Chunk) erro
...
+	}
+	defer terror.Call(result.Close)
+
+	sc := e.ctx.GetSessionVars().StmtCtx
+	for {
+		err := result.NextChunk(ctx, e.idxChunk)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if e.idxChunk.NumRows() == 0 {
```
[the remark this answers is not in the document]
Comment (zhexuany) [reply executor/admin.go]: friendly ping @jackysp Please resolve reviewer's comment

**Comment by jackysp** [comment]
PTAL @zimulala

File: executor/admin.go [hunk 6]
```diff
@@ -457,3 +459,248 @@ func (e *RecoverIndexExec) NextChunk(ctx context.Context, chk *chunk.Chunk) erro
...
+			if e.removeCnt%e.batchSize == 0 {
+				log.Infof("[cleaning up dangling index] table: %v, index: %v, count: %v.",
+					e.table.Meta().Name.String(), e.index.Meta().Name.String(), e.removeCnt)
+			}
+		}
+	}
+	return nil
+}
+
+func (e *CleanupIndexExec) extractIdxVals(row chunk.Row, idxVals []types.Datum) []types.Datum {
```
Comment (zimulala) [inline executor/admin.go]: This function logic is the same as RecoverIndexExec's extractIdxVals. Could we use the same one?

File: executor/admin.go [hunk 7]
```diff
@@ -457,3 +459,248 @@ func (e *RecoverIndexExec) NextChunk(ctx context.Context, chk *chunk.Chunk) erro
...
+		}
+		if e.idxChunk.NumRows() == 0 {
+			return nil
+		}
+		iter := chunk.NewIterator4Chunk(e.idxChunk)
+		for row := iter.Begin(); row != iter.End(); row = iter.Next() {
+			handle := row.GetInt64(len(e.idxCols) - 1)
+			idxVals := e.extractIdxVals(row, e.idxValsBufs[e.scanRowCnt])
+			e.idxValsBufs[e.scanRowCnt] = idxVals
+			e.matchedIndex[handle] = idxData{false, idxVals}
```
Comment (zimulala) [inline executor/admin.go]: Could we remove the field of`matched `?

File: executor/admin.go [hunk 8]
```diff
@@ -457,3 +459,248 @@ func (e *RecoverIndexExec) NextChunk(ctx context.Context, chk *chunk.Chunk) erro
...
+	for i := range e.idxCols {
+		dagReq.OutputOffsets = append(dagReq.OutputOffsets, uint32(i))
+	}
+
+	execPB := e.constructIndexScanPB()
+	dagReq.Executors = append(dagReq.Executors, execPB)
+	err := plan.SetPBColumnsDefaultValue(e.ctx, dagReq.Executors[0].IdxScan.Columns, e.idxCols)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
```
Comment (zimulala) [inline executor/admin.go]: Can these codes be extracted into a function？
The logic is used in many places.

Comment (jackysp) [reply executor/admin.go]: They looked similar, but not exactly the same. It needs a refactor. Maybe in the next pr.

**Comment by jackysp** [comment]
PTAL @zimulala

File: executor/builder.go [hunk 9]
```diff
@@ -295,6 +297,52 @@ func (b *executorBuilder) buildRecoverIndex(v *plan.RecoverIndex) Executor {
...
+
+func (b *executorBuilder) buildCleanupIndex(v *plan.CleanupIndex) Executor {
+	tblInfo := v.Table.TableInfo
+	t, err := b.is.TableByName(v.Table.Schema, tblInfo.Name)
+	if err != nil {
+		b.err = errors.Trace(err)
+		return nil
+	}
+	idxName := strings.ToLower(v.IndexName)
+	indices := t.WritableIndices()
```
Comment (zimulala) [inline executor/builder.go]: Is it enough if we only handle the public index?

File: executor/admin.go (hunk 8 above)
Comment (zimulala) [reply executor/admin.go]: It's the same as `CheckIndexRangeExec`'s buildDAGPB.
But if you want to refactor some of the exec in admin, you can handle it in the next PR.

File: executor/builder.go [hunk 10]
```diff
@@ -320,9 +320,8 @@ func (b *executorBuilder) buildCleanupIndex(v *plan.CleanupIndex) Executor {
 		return nil
 	}
 	idxName := strings.ToLower(v.IndexName)
-	indices := t.WritableIndices()
 	var index table.Index
-	for _, idx := range indices {
```
Comment (zimulala) [inline executor/builder.go]: `Indices()` isn't equal to public indices.

**Comment by jackysp** [comment]
PTAL @zimulala , sorry for the late update.

File: executor/builder.go [hunk 11]
```diff
@@ -321,7 +321,10 @@ func (b *executorBuilder) buildCleanupIndex(v *plan.CleanupIndex) Executor {
 	}
 	idxName := strings.ToLower(v.IndexName)
 	var index table.Index
-	for _, idx := range t.WritableIndices() {
+	for _, idx := range t.Indices() {
+		if idx.Meta().State != model.StatePublic {
```
Comment (zimulala) [inline executor/builder.go]: I think to do this check as line 334 is better.

**Pull request merged**